transport-native-epoll/src/main/c/netty_epoll_native.c

```diff
@@ -404,92 +406,221 @@ static jint netty_epoll_native_tcpMd5SigMaxKeyLen(JNIEnv* env, jclass clazz) {
+               goto error;
+            }
+
+            io_prep_pread(iocbp, fd, (void *)buf_address, length, offset);
+        }
+        else {
+            if (nbuffers <= 0 || nbuffers > MAX_NUM_BUFFERS_PER_REQUEST) {
+                netty_unix_errors_throwRuntimeException(env, "invalid number of buffers, should be positive and less than MAX_NUM_BUFFERS_PER_REQUEST, typically 8.");
+               goto error;
+            }
+
+            //printf("Processing vectored request with %d buffers\n", nbuffers);
+
+            for(j = 0; j < nbuffers; j++) {
+                long buf_address = buf_addresses[nbuffers_processed + j];
+                long length = buf_lengths[nbuffers_processed + j];
+
+                //printf("...buffer %ld of length %ld\n", buf_address, length);
+
+                if (buf_address & 511 != 0) {
+                    netty_unix_errors_throwRuntimeException(env, "buffer is not memory aligned");
+                    goto error;
+                }
+
+                niocbp->iovec[j].iov_base = (void *)buf_address;
+                niocbp->iovec[j].iov_len = length;
+            }
+
+            io_prep_preadv(iocbp, fd, niocbp->iovec, nbuffers, offset);
+        }
+
+        nbuffers_processed += nbuffers;
+        io_set_eventfd(iocbp, efd);
+    }
+
+    //printf("Submitting request\n");
+    int r = io_submit(ctx->aio, num_requests, iocbps);
+    if (r != num_requests) {
+        //printf("io_submit() failed with %d\n", r);
         netty_unix_errors_throwChannelExceptionErrorNo(env, "io_submit() failed: ", -r);
+        goto error;
+    }
+
+    goto cleanup;
+
+error: // release the slots
+    for(i = 0; i < num_requests; i++) {
+        netty_iocb_t* niocbp = (netty_iocb_t *)iocbps[i];
+        if (niocbp != NULL) {
+            niocbp->slot = -1;
+        }
     }
+
+cleanup: // release temporary arrays
+
+    free(iocbps); // can be freed after io_submit()
+
+    (*env)->ReleaseIntArrayElements(env, in_slots, slots, 0);
+    (*env)->ReleaseLongArrayElements(env, in_offsets, offsets, 0);
+    (*env)->ReleaseIntArrayElements(env, in_num_buffers, num_buffers, 0);
+    (*env)->ReleaseLongArrayElements(env, in_buf_addresses, buf_addresses, 0);
+    (*env)->ReleaseLongArrayElements(env, in_buf_lengths, buf_lengths, 0);
 }
 
-static void netty_epoll_native_getAIOEvents0(JNIEnv* env, jclass clazz, jlong ctxaddress, jlong num_events, jlongArray keys)
-{
-    io_context_t *ctx = (io_context_t *) ctxaddress;
-    struct io_event events[num_events];
-    unsigned long keysArray[num_events * 2];
+static jint netty_epoll_native_getAIOEvents0(JNIEnv* env, jclass clazz, jlong ctxaddr, jlongArray result) {
+
+    netty_io_context_t* ctx = (netty_io_context_t*) ctxaddr;
+    struct io_event events[ctx->concurrency];
+
+    struct timespec timeout;
+    timeout.tv_sec = 0;
+    timeout.tv_nsec = 0;
+
+    int r, j, slot;
+    r = io_getevents(ctx->aio, 1, ctx->concurrency, events, &timeout);
```

Comment:
Don't we know from epoll exactly how many events are ready? Shouldn't we wait until we read that many?

Reply:
We save a call to `eventfd_read` by reading up to concurrency. It's fine at the moment because the concurrency is small, see point above w.r.t. revisiting this for DB-1463. The timeout is zero, so even if there are no events it returns immediately.